src/charm.py

```diff
@@ -0,0 +1,226 @@
+        elif not self._loki_cluster.relation:
+            e.add_status(WaitingStatus("Loki-Cluster relation not ready"))
+        if not self._loki_cluster.get_loki_config():
+            e.add_status(WaitingStatus("Waiting for coordinator to publish a loki config"))
+        if not self._loki_roles:
+            e.add_status(BlockedStatus("No roles assigned: please configure some roles"))
+        e.add_status(ActiveStatus(""))
+
+    def _on_loki_cluster_created(self, _):
+        self._update_loki_cluster()
+
+    def _on_loki_config_received(self, _e: ConfigReceivedEvent):
+        self._update_config()
+
+    # === PROPERTIES === #
+
+    @property
+    def _loki_roles(self) -> List[LokiRole]:
+        """Return a set of the roles this Loki worker should take on."""
+        roles: List[LokiRole] = [role for role in LokiRole if self.config[role] is True]
+        return roles
+
+    @property
+    def server_cert_path(self) -> Optional[str]:
+        """Server certificate path for tls tracing."""
+        return LOKI_CERT_FILE
+
+    # === UTILITY METHODS === #
+
+    def _update_loki_cluster(self):
+        """Share via loki-cluster all information we need to publish."""
+        self._loki_cluster.publish_unit_address(socket.getfqdn())
+        if self.unit.is_leader() and self._loki_roles:
+            logger.info(f"Publishing Loki roles: {self._loki_roles} to relation databag.")
+            self._loki_cluster.publish_app_roles(self._loki_roles)
+
+    def _update_config(self):
+        """Update the loki config and restart the workload if necessary."""
+        cert_secrets = self._loki_cluster.get_cert_secret_ids()
+        loki_config = self._loki_cluster.get_loki_config()
+        restart = any(
+            [
+                self._loki.update_tls_certificates(  # pyright: ignore[reportOptionalMemberAccess]
+                    cert_secrets
+                ),
+                self._loki.update_config(  # pyright: ignore[reportOptionalMemberAccess]
+                    loki_config
+                ),
+                self._set_pebble_layer(),
+            ]
+        )
+
+        if restart:
+            self._loki.restart()  # pyright: ignore[reportOptionalMemberAccess]
+
+    def _set_pebble_layer(self) -> bool:
+        """Set Pebble layer.
+
+        Returns: True if Pebble layer was added, otherwise False.
+        """
+        if not self._container.can_connect():
+            return False
+        if not self._loki_roles:
+            return False
+
+        current_layer = self._container.get_plan()
+        new_layer = self._loki.pebble_layer  # pyright: ignore[reportOptionalMemberAccess]
+
+        if (
+            "services" not in current_layer.to_dict()
+            or current_layer.services != new_layer["services"]
+        ):
+            self._container.add_layer(CONTAINER_NAME, new_layer, combine=True)  # pyright: ignore
+            return True
+
+        return False
+
+
+class ManualLogForwarder(Object):
```

Comment:
Is this something that needs to go into the loki push api lib?
I.e. why do we roll a new one here?

Reply:
@lucabello do you remember the rationale of this custom class?
We implemented it in [mimir-worker for the first time](https://github.com/canonical/mimir-worker-k8s-operator/blame/main/src/charm.py#L324)

Feels like we can generalise it and include it in the lib...